modules/generator/encoding.go

```diff
@@ -0,0 +1,63 @@
+package generator
+
+import (
+	"iter"
+
+	"github.com/grafana/tempo/pkg/ingest"
+	"github.com/grafana/tempo/pkg/tempopb"
+)
+
+// codec is the interface used to convert data from Kafka records to the
+// tempopb.PushSpansRequest expected by the generator processors.
+type codec interface {
+	decode([]byte) (iter.Seq2[*tempopb.PushSpansRequest, error], error)
+}
+
+// tempoDecoder unmarshals tempopb.PushBytesRequest.
+type tempoDecoder struct {
+	dec *ingest.Decoder
+}
+
+func newTempoDecoder() *tempoDecoder {
+	return &tempoDecoder{dec: ingest.NewDecoder()}
+}
+
+// decode implements codec.
+func (d *tempoDecoder) decode(data []byte) (iter.Seq2[*tempopb.PushSpansRequest, error], error) {
+	d.dec.Reset()
+	spanBytes, err := d.dec.Decode(data)
+	if err != nil {
+		return nil, err
+	}
+
+	trace := tempopb.Trace{}
+	return func(yield func(*tempopb.PushSpansRequest, error) bool) {
+		for _, tr := range spanBytes.Traces {
+			trace.Reset()
+			err = trace.Unmarshal(tr.Slice)
+			yield(&tempopb.PushSpansRequest{Batches: trace.ResourceSpans}, err)
+		}
+	}, nil
+}
+
+// otlpDecoder unmarshals ptrace.Traces.
+type otlpDecoder struct {
+	trace tempopb.Trace
+}
+
+func newOTLPDecoder() *otlpDecoder {
+	return &otlpDecoder{trace: tempopb.Trace{}}
+}
+
+// decode implements codec.
+func (d *otlpDecoder) decode(data []byte) (iter.Seq2[*tempopb.PushSpansRequest, error], error) {
+	d.trace.ResourceSpans = d.trace.ResourceSpans[:0]
+	err := d.trace.Unmarshal(data)
+	if err != nil {
+		return nil, err
+	}
+
+	return func(yield func(*tempopb.PushSpansRequest, error) bool) {
+		yield(&tempopb.PushSpansRequest{Batches: d.trace.ResourceSpans}, nil)
```

Comment:
so we are returning a slice owned and reused by this struct into the bowels of the metrics generator. we are certain that we will not reuse it here while it's being processed in the generator?

Reply:
Pretty certain from what I can discern. This is also following the pattern of what the code is already doing before this PR, so I think it should be fine.